runtime/src/main/kotlin/kotlin/native/concurrent/SharedRef.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package kotlin.native.concurrent
+
+import kotlin.native.internal.*
+
+@SymbolName("Kotlin_SharedRef_createSharedRef")
+external private fun createSharedRef(value: Any): NativePtr
+
+/**
+ * A frozen shared reference to a Kotlin object
+ *
+ * Can be safely passed between workers, but can only be accessed on the worker it was created on
+ *
+ * Note: Garbage collector currently cannot detect reference cycles between [SharedRef]s. Consider
+ * using [DisposableSharedRef] to manually resolve cycles with explicit calls to [DisposableSharedRef.dispose]
+ */
+@Frozen
+@NoReorderFields
+@ExportTypeInfo("theSharedRefTypeInfo")
+public class SharedRef<out T : Any> private constructor(private var ptr: NativePtr) {
+
+    companion object {
+        /**
+         * Creates shared reference to given object
+         */
+        fun <T : Any> create(value: T) = SharedRef<T>(createSharedRef(value))
+    }
+
+    /**
+     * Returns the object this reference was [created][SharedRef.create] for.
+     */
+    @SymbolName("Kotlin_SharedRef_derefSharedRef")
+    external fun get(): T
```

Comment:
Note that we have an option to make this a property (`.value`) with custom getter. This would be consistent with other classes in K/N (consider `AtomicReference`). But I'm not sure that property getter should throw an exception for illegal access.
A piece of advice from @ilya-g would be extremely useful here.

Reply:
It meets two of three criteria when property should be preferred: https://kotlinlang.org/docs/reference/coding-conventions.html#functions-vs-properties

Sometimes, the first one ("Does not throw") can be dismissed, for example `Lazy.value` is a property that can throw whatever the lazy initializer throws. Whether or not it is acceptable depends on the use case.